src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,467 @@
+    }
+
+    // --- Allow DAI holders to exit during global settlement ---
+    function exit(bytes32 ilk_, address usr, uint256 wad) external {
+        require(wad <= 2 ** 255, "DssDirectDepositHub/overflow");
+        vat.slip(ilk_, msg.sender, -int256(wad));
+        D3MPoolLike pool = ilks[ilk_].pool;
+        require(pool.transferShares(usr, pool.convertToShares(wad)), "DssDirectDepositHub/failed-transfer");
+        emit Exit(ilk_, usr, wad);
+    }
+
+    // --- Shutdown ---
+    function cage(bytes32 ilk_) external {
+        require(vat.live() == 1, "DssDirectDepositHub/no-cage-during-shutdown");
+
+        D3MPoolLike pool = ilks[ilk_].pool;
+
+        // Can shut joins down if we are authed
+        // or if the interest rate strategy changes
+        // or if the main module is caged
+        require(
+            wards[msg.sender] == 1 ||
+            live == 0 ||
+            !pool.validTarget()
+        , "DssDirectDepositHub/not-authorized");
+
+        pool.cage();
+        ilks[ilk_].tic = block.timestamp;
+        emit Cage(ilk_);
+    }
+
+    function cage() external {
+        require(wards[msg.sender] == 1 , "DssDirectDepositHub/not-authorized");
+        require(vat.live() == 1, "DssDirectDepositHub/no-cage-during-shutdown");
+
+        live = 0;
+        emit Cage();
+    }
+
+    // --- Write-off ---
+    function cull(bytes32 ilk_) external {
+        require(vat.live() == 1, "DssDirectDepositHub/no-cull-during-shutdown");
+        require(live == 0, "DssDirectDepositHub/live");
+        Ilk storage ilk = ilks[ilk_];
+        require(ilk.tic > 0, "DssDirectDepositHub/join-live");
+        require(_add(ilk.tic, ilk.tau) <= block.timestamp || wards[msg.sender] == 1, "DssDirectDepositHub/unauthorized-cull");
+        require(ilk.culled == 0, "DssDirectDepositHub/already-culled");
+
+        (uint256 ink, uint256 art) = vat.urns(ilk_, address(ilk.pool));
+        require(ink <= 2 ** 255, "DssDirectDepositHub/overflow");
+        require(art <= 2 ** 255, "DssDirectDepositHub/overflow");
+        vat.grab(ilk_, address(ilk.pool), address(ilk.pool), vow, -int256(ink), -int256(art));
+
+        ilk.culled = 1;
+        emit Cull(ilk_);
+    }
+
+    // --- Rollback Write-off (only if General Shutdown happened) ---
+    // This function is required to have the collateral back in the vault so it can be taken by End module
+    // and eventually be shared to DAI holders (as any other collateral) or maybe even unwinded
+    function uncull(bytes32 ilk_) external {
+        D3MPoolLike pool = ilks[ilk_].pool;
+
+        require(ilks[ilk_].culled == 1, "DssDirectDepositHub/not-prev-culled");
+        require(vat.live() == 0, "DssDirectDepositHub/no-uncull-normal-operation");
+
+        uint256 wad = vat.gem(ilk_, address(pool));
+        require(wad < 2 ** 255, "DssDirectDepositHub/overflow");
+        vat.suck(vow, vow, _mul(wad, RAY)); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
```

Comment:
Not big deal as not common function, but it should be more gas efficient to save `vow` in a memory variable and use that.